### PR TITLE
:eight: Lecture 8: ZK-SNARK, Arithmetic Circuits, R1CS, QAP

### PR DESCRIPTION
**Pull request by OmegaTymbJIep**
# Objective

* Introduction to the ZK-SNARK subject
* Arithmetic circuits
* R1CS
* QAP

### CODE REVIEW

File: lectures/zksnark.tex [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+\subsection{What the zk-SNARK is?}
+
+Let's first discuss what zk-SNARK is.
+
+\begin{definition}
+    \textbf{zk-SNARK} – Zero-Knowledge Succinct Non-interactive ARgument of Knowledge.
+\end{definition}
+
+But what do terms like "argument of knowledge," "succinct," "non-interactive," and
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: You should use ``Something here'' for quotes

File: lectures/zksnark.tex [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    \textbf{zk-SNARK} – Zero-Knowledge Succinct Non-interactive ARgument of Knowledge.
+\end{definition}
+
+But what do terms like "argument of knowledge," "succinct," "non-interactive," and 
+"zero-knowledge" mean in this context?
+
+\begin{itemize}
+    \item \textbf{Argument of Knowledge} - a proof that the prover know data that resolves a certain
+    problem, and this knowledge can be verified.
+    \item \textbf{Succinct} - the proof size is relatively small and does not depend on the size of
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Specify that you will show what "relatively small" means specifically in the further discussion

File: lectures/zksnark.tex [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+any information about the value itself, and do so with a proof that is both small and quick to 
+verify. This makes zk-SNARKs a powerful tool for maintaining privacy and efficiency in various 
+cryptographic applications.
+
+This is pretty wide defined and maybe not so obvious if you don't have any background, so let's take
+a look at the example.
+
+\begin{example}
+    
+    Imagine you're part of a treasure hunt, and you've found a hidden treasure chest. You want to
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: I like the example!

File: lectures/zksnark.tex [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+resolves a specific challenge or gives us a particular result.
+
+\subsection{Arithmetic Circuits}
+The cryptographic tools we've learned in the previous sections work with numbers or certain 
+primitives above them, so the first question is: how do we convert a program into mathematical 
+language? Additionally, we need to do this in a way that is succinct and allows us to prove 
+something about it.
+
+The \textbf{Arithmetic circuits} can help us. Similar to boolean circuits, they have gates and 
+wires, but instead of the operations $AND$, $OR$, and $NOT$, only multiplication and addition are
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: I propose to change the font as follows: `\mathsf{AND}`

File: lectures/zksnark.tex [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        \hline
+    \end{tabular}
+    \caption{AND Gate Truth Table}
+    \label{tab:and-truth-table}
+    \vspace{1em}
+\end{wraptable}
+
+% --- Finish Writing diagrams ---
+
+The \textit{AND Gate Truth Table~\ref{tab:and-truth-table}} shows us the results we receive if
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Instead of `Table~\ref{...}` you can instead use `Cref{...}` using the [`cleveref`](https://ctan.org/pkg/cleveref?lang=en) package

File: lectures/zksnark.tex [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+The witness vector (solution vector) is $w = \{r, a, b\}$, for example: $\{6, 2, 3\}$. We 
+assume that the $a$ and $b$ are input values. 
+
+We can think of the "=" in the gate as an assertion, meaning that if $a \times b$ does not equal
+$r$, the assertion fails, and the input values don't resolve the circuit.
+
+How can we translate an if statement?
+
+\begin{lstlisting}[numbers=none]
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Probably specify some language (like Python or Rust) to add highlighting to expressions

File: lectures/zksnark.tex [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+We can express this logic in mathematical terms as follows: "If $a$ is true, compute 
+$b \times c$; otherwise, compute $b + c$." Only numerical expressions are allowed. Knowing that
+$\text{true} := 1$ and $\text{false} := 0$, we can transform it as follows:
+
+\begin{equation}
+    r = a \times (b \times c) + (1 - a) \times (b + c)
+\end{equation}
+
+$w = \{r, a , b, c\}$: $\{6, 1, 2, 3\}$, $\{5, 0, 2, 3\}$. \\
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Specify, "It seems that the witness $w$ should be of a form..."

Besides, I believe it is better to use a tuple notation to represent a witness (that is $(r, a,b,c)$ instead of $\\{r,a,b,c\\}$). After all, the witness is not a set but an ordered set of scalars (a vector, in a sense).

**Review by ZamDimon — CHANGES_REQUESTED** [review]
A solid starting point!

File: lectures/zksnark.tex [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    \end{aligned}
+\end{equation*}
+
+With witness vector: $w = (r, r_1, r_2, r_3, r_4, r_5, a , b, c)$: $(6, 6, 5, 0, 6, 0, 1, 2, 3)$.
+
+\subsection{Rank-1 Constraint System}
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constrain System}, where the simpliest term is \textbf{constraint}.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Constrain**t**

File: lectures/zksnark.tex [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constrain System}, where the simpliest term is \textbf{constraint}. 
+This offers a more flexible and general way to describe these parts.
+    
+\begin{definition}
+    Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An \textbf{inner product} 
+    on $\mathbb{V}$ is a function:
+    \[ \<\cdot, \cdot \>: \mathbb{V} \times \mathbb{V} \rightarrow \mathbb{F} \]
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Just use `\langle` and `\rangle` for brackets $\langle \cdot \rangle$

File: lectures/zksnark.tex [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+    defined for $u, v \in \mathbb{V}$ as a linear combination:
+    \[ \< u, v \> = \sum_{i=1}^{n} u_i v_i \]
+    With next properties:
+    \begin{itemize}
+        \item Symmetry: $\forall u, v \in \mathbb{V}: \< u, v \> = \overline{\< v, u \>}$
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Mention that this "overline" means conjugation and is typically relevant when $\mathbb{F} = \mathbb{C}$

File: lectures/zksnark.tex [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        \item Positivity: $\forall u \in \mathbb{V}: \< u, u \> \geq 0$ and 
+        $\< u, u \> = 0 \Leftrightarrow u = 0$
+    \end{itemize}
+\end{definition}
+
+
+\begin{example}
+    Let $u, v$ are vectors over the real number $\mathbb{R}$, where
+    \[u = (1, 2, 3), v = (2, 4, 3)\], then:
+    $\<u, v\> = \sum_{i=1}^{3}u_iv_i = 2*1 + 2*4 + 3*3 = 2 + 8 + 9 = 19$
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Either use `\cdot` or `\times` for denoting multiplication. Since `*` is typically used to represent the convolution operation

File: lectures/zksnark.tex [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+\end{example}
+
+With knowledge of the inner product of two vectors, we can now formulate a definition of the 
+constraint in the context of an R1CS.
+
+\begin{definition}
+    Each \textbf{constraint} in the Rank-1 Constraint System must be in the form:
+    \[ \<a, w\> \times \<b, w\> = \<c, w\> \]
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: If you want formula without numeration, better to use `\begin{equation*}...\end{equation*}`

File: lectures/zksnark.tex [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+    Where $\mathbf{w}$ is a vector containing all the \textit{input}, \textit{output}, and
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Make notation consistent. If you do not make any special notation for a vector (like, $w$), do not use `\mathbf{...}` here. Otherwise, use this `\mathbf{...}` ($\mathbf{w}$) everywhere (which I do recommend). Alternatively, quite frequently, one denotes vectors using `\vec{...}` to put an arrow above ($\vec{w}$).

File: lectures/zksnark.tex [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+
+    However, one important consideration was overlooked. If $x_1$ is neither $0$ nor $1$, it implies
+    that something else is being computed instead of the desired program. Since we need to add a
+    restriction for $x_1$: $x_1 \times (1 - x_1) = 0$, this effectively checks that $x_1$ is binary.
+
+    The next constraints can be build:
+    \begin{align*}
+        x_1 \times x_1 &= x_1 \quad \text{(binary check)} \tag{1} \\
+        x_2 \times x_3 &= mult \tag{2} \\
+        x_1 \times mult &= selectMult \tag{3} \\
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Either do `\mathsf{mult}` or use `t_1` or `\mathsf{temp}_1` to denote intermediate temporary variables

File: lectures/zksnark.tex [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+    \[ w = (1, r, x_1, x_2, x_3, mult, selectMult) \]
+    The coefficients vectors:
+    \begin{align*}
+        a_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad c_1 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_2 &= (0, 0, 0, 1, 0, 0, 0) & \quad b_2 &= (0, 0, 0, 0, 1, 0, 0) & \quad c_2 &= (0, 0, 0, 0, 0, 1, 0) \\
+        a_3 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_3 &= (0, 0, 0, 0, 0, 1, 0) & \quad c_3 &= (0, 0, 0, 0, 0, 0, 1) \\
+        a_4 &= (1, 0, -1, 0, 0, 0, 0) & \quad b_4 &= (0, 0, 0, 1, 1, 0, 0) & \quad c_4 &= (0, 1, 0, 0, 0, 0, -1)
+    \end{align*}
+
+    Now, let's use specific values to compute an example:
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: This example is too long, try splitting it in two

File: lectures/zksnark.tex [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+    \[ x_1 = 1, \quad x_2 = 3, \quad x_3 = 4 \]
+
+    We get:
+    \[ \text{mult} = x_2 \times x_3 = 3 \times 4 = 12 \]
+    \[ \text{selectMult} = x_1 \times \text{mult} = 1 \times 12 = 12 \]
+    \[ r = x_1 \times (x_2 \times x_3) + (1 - x_1) \times (x_2 + x_3) \]
+    \[ r = 1 \times (3 \times 4) + (1 - 1) \times (3 + 4) = 12 \]
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Maybe I missed that, but specify somewhere that you are using $\mathbb{R}$ (or some large finite field $\mathbb{F}_p$) for all the arithmetic

File: lectures/zksnark.tex [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+the defined relationships, thus encoding the entire computational process in the form of a system
+of rank-1 quadratic equations.
+
+The last unresolved question is where the ``rank-1`` comes from.
+
+\begin{lemma}
+    Suppose there is a constraint $\<a, w\> \times \<b, w\> = \< c, w \>$ with coefficient vectors
+    $a$, $b$, $c$ and witness vector $w$. Then it can be expressed in the form:
+    \[ w^T A w + c^T w = 0 \]
+    Where $A$ is the outer product of vectors $a$, $b$ (denoted as $a \otimes b$), consequently
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Describe what the outer product is (since only the inner product was defined previously).

Additionally, it could make sense to make a separate subsection for "Linear Algebruh Preliminaries", but it is up to you to decide

File: lectures/zksnark.tex [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+\begin{lemma}
+    Suppose there is a constraint $\<a, w\> \times \<b, w\> = \< c, w \>$ with coefficient vectors
+    $a$, $b$, $c$ and witness vector $w$. Then it can be expressed in the form:
+    \[ w^T A w + c^T w = 0 \]
+    Where $A$ is the outer product of vectors $a$, $b$ (denoted as $a \otimes b$), consequently 
+    a \textbf{rank-1} matrix.
+\end{lemma}
+
+\textbf{Lemma proof.} Suppose there is a constraint $\<a, w\> \times \<b, w\> = \<c, w\>$. Where
+vectors $a$, $b$, $c$, and $w$ have length $n$. Let's expand the inner products:
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: This condition formulation can be simplified to $a,b,c,w \in \mathbb{F}^n$

**Review by ZamDimon — COMMENTED** [review]
Everything looks **very** solid to me; keep it up!

File: lectures/zksnark.tex [hunk 19]
```diff
@@ -0,0 +1,812 @@
...
+
+\subsection{What the zk-SNARK is?}
+
+Let's first discuss what zk-SNARK is.
+
+\begin{definition}
+    \textbf{zk-SNARK} – Zero-Knowledge Succinct Non-interactive ARgument of Knowledge.
+\end{definition}
+
+But what do terms like ``argument of knowledge``, ``succinct``, ``non-interactive``, and
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: I guess after the "Introduction to Zero-Knowledge" lecture, we might take a more formal view of these terms (I might myself add a couple of lines here). Nonetheless, we still need to maintain simplistic explanations of what that is for better general understanding!

File: lectures/zksnark.tex [hunk 20]
```diff
@@ -0,0 +1,812 @@
...
+        \item Zero-Knowledge: The proof doesn't reveal any information about the actual location of
+        the treasure chest. The organizer knows you found it, but they don't learn anything about 
+        where it is hidden. \\
+    \end{itemize}
+
+    Here you can think of zk-SNARK as a golden coin from the chest where the pirates' sign is 
+    engraved, so the organizer can be sure you've found the treasure.
+\end{example}
+
+But the problems that we want to solve are in a slightly different format. We can't bring a coin to
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: This example aligns with our previous lecture very well!

File: lectures/zksnark.tex [hunk 21]
```diff
@@ -0,0 +1,812 @@
...
+    \end{aligned}
+\end{equation*}
+
+With witness vector: $w = (r, r_1, r_2, r_3, r_4, r_5, a , b, c)$: $(6, 6, 5, 0, 6, 0, 1, 2, 3)$.
+
+\subsection{Rank-1 Constraint System}
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constraint System}, where the simpliest term is \textbf{constraint}.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Long dashes should be written either as `---` or `--`.

File: lectures/zksnark.tex [hunk 22]
```diff
@@ -0,0 +1,812 @@
...
+
+\subsection{Rank-1 Constraint System}
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constraint System}, where the simpliest term is \textbf{constraint}. 
+This offers a more flexible and general way to describe these parts.
+    
+\begin{definition}
+    Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An \textbf{inner product}
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Well, I would rephrase this definition a bit. 

The inner product of a linear space $\mathbb{V}$ is **any** symmetric, linear in the first argument, and positive binary function from vector space to a set of scalars. The specific function we are used to $\langle u,v \rangle = \sum_{i=1}^n u_iv_i$ (which is formally not a linear combination) is just a special case of this general definition. You wrote everything right, but formally the properties should go before the specific function we are using in SNARKs.

Comment (ZamDimon) [reply lectures/zksnark.tex]: One example of another valid scalar product formulation for $\mathbb{R}^n$ is $\langle u,v \rangle := \sum_{i=1}^n \mu_iu_iv_i$ for some positive set $\\{\mu_i\\}_{i=1}^n \subset \mathbb{R}\_{>0}$.

Moreover, any symmetric positive-definite matrix $M$ would make $\langle u,v \rangle = u^{\top}Mv$ a valid scalar product! :) But that is just for you to know, no need to insert it to make matters more complicated

File: lectures/zksnark.tex [hunk 23]
```diff
@@ -0,0 +1,812 @@
...
+
+Thus, the constraint can be written as:
+\[ w^T A w + c^T w = 0 \]
+
+So, the rank-1 means the rank of the coefficients matrix $A$ in one of the constraint formats.
+
+\subsection{Quadratic Arithmetic Program}
+
+While the Rank-1 Constraint System provides a powerful way to represent computations, it is not 
+succinct at all, since the number of constraints directly depends on the complexity of the problem
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: "Directly depends" is a bit vague term. I believe it would be appropriate to say that it "depends linearly on the complexity of the problem being solved."

File: lectures/zksnark.tex [hunk 24]
```diff
@@ -0,0 +1,812 @@
...
+being solved. In practical scenarios, this can require tens or even hundreds of thousands of 
+constraints, sometimes even millions. The Quadratic Arithmetic Program (QAP) can address this issue.
+
+\begin{remark}
+    Understanding polynomials and their properties is crucial for this section. If you are not 
+    confident in this area, it is better to revisit the corresponding chapter and refresh your
+    knowledge.
+\end{remark}
+
+To define a constraint in the R1CS we need four vectors: three coefficient vectors ($a$, $b$, and
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Let us denote vectors either by `\vec{...}` ($\vec{a}$) or `\mathbf{...}` ($\mathbf{a}$) to distinguish them from the scalars.

File: lectures/zksnark.tex [hunk 25]
```diff
@@ -0,0 +1,812 @@
...
+$c$) and the witness one ($w$). And that's just for one constraint. As you can imagine, many of
+the values in these vectors are zeros. In circuits with thousands of inputs, outputs, and auxiliary
+variables, where there are also thousands of constraints, you could end up with millions of zeroes.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: (in numerical analysis, such vectors and matrices are typically called "sparse" (https://en.wikipedia.org/wiki/Sparse_matrix))

File: lectures/zksnark.tex [hunk 26]
```diff
@@ -0,0 +1,812 @@
...
+            c_{11} & c_{12} & \dots & c_{1n} \\
+            c_{21} & c_{22} & \dots & c_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            c_{m1} & c_{m2} & \dots & c_{mn}
+        \end{pmatrix}
+    \end{align*}
+    
+    In this representation:
+    \begin{itemize}
+        \item Each row of the matrices corresponds to the coefficients of a specific constraint.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: For better understanding, I would rephrase it as "Each i-th row of the matrix corresponds to the corresponding coefficients in the i-th constraint".

File: lectures/zksnark.tex [hunk 27]
```diff
@@ -0,0 +1,812 @@
...
+        \item Each column of these matrices corresponds to the coefficients associated with a 
+        particular element of the witness vector $w$.
+    \end{itemize}
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: More formally, if your R1CS constraints consist of vectors $\\{a_i\\}_{i=1}^m$, then the matrix $A$ can be represented as:

$$
A = \begin{bmatrix}
a_1^{\top} \\
a_2^{\top} \\
\vdots \\
a_m^{\top}
\end{bmatrix}
$$

File: lectures/zksnark.tex [hunk 28]
```diff
@@ -0,0 +1,812 @@
...
+
+The columns of these matrices represent the mappings from constraint number $i$ to the corresponding
+coefficient of the $j$ element in the witness vector.
+
+\begin{example}
+    Considering the witness from the previos examples: 
+    \begin{equation*}
+         w = (1, r, x_1, x_2, x_3, \mathsf{mult}, \mathsf{selectMult})
+    \end{equation*}
+    For element $x_1$ we are interesting in the third columns of the $A$, $B$ and $C$ matrices, as
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: For element $x_1$, we are interest**ed** in the third columns of $A,B,C$ matrices

File: lectures/zksnark.tex [hunk 29]
```diff
@@ -0,0 +1,812 @@
...
+
+\begin{example}
+    The vectors $a_i$ from the previous examples:
+    \begin{align*}
+        a_1 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_2 &= (0, 0, 0, 1, 0, 0, 0) \\
+        a_3 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_4 &= (1, 0, -1, 0, 0, 0, 0)
+    \end{align*}
+    $n = 7, m = 4$
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: This corresponds to $n=7,m=4$

File: lectures/zksnark.tex [hunk 30]
```diff
@@ -0,0 +1,812 @@
...
+        
+        \end{tikzpicture}
+    \end{center}
+\end{example}
+
+As we know from the previous chapters, such a mapping in math can be built using polynomial 
+interpolation.
+
+\begin{remark}
+    As a remainder the Lagrange interpolation polynomial for a given set of points
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Comma after "as a remainder"

File: lectures/zksnark.tex [hunk 31]
```diff
@@ -0,0 +1,812 @@
...
+    As a remainder the Lagrange interpolation polynomial for a given set of points 
+    $\{(x_0,y_0),(x_1,y_1),\dots,(x_n,y_n)\} \subset \mathbb{F} \times \mathbb{F}$
+    can be built with the following formula:
+    \begin{equation*}
+        L(x) = \sum_{i=0}^{n} y_i \ell_i(x), \quad \ell_i(x) = \prod_{j=0, j \neq i}^{n} \frac{x-x_j}{x_i-x_j}.
+    \end{equation*}  
+\end{remark}
+
+For a given column $j \in \{1, 2, \dots, n\}$ in a matrix $A$ the set of points that define the
+variable polynomial $a_j(x)$ can be defined as follows:
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Use some other letter for representing polynomials (since you used $a$ multiple times before). Maybe capital $A_j(x)$ or $P^A_j(x)$ or $L_j(x)$ (corresponding to **left** input writes to the gates)?

File: lectures/zksnark.tex [hunk 32]
```diff
@@ -0,0 +1,812 @@
...
+        \node at (axis cs:3,1.5) [anchor=west] {\small (3,1)};
+        \node at (axis cs:4,-1) [anchor=west] {\small (4,-1)};
+        
+        \end{axis}
+    \end{tikzpicture}
+    \caption{The Lagrange inteprolation polynomial for points $\{(1,1), (2,0), (3,1), (4,-1)\}$}
+    \label{fig:example-interpolation}
+\end{figure}
+
+The degree of coefficient polynomials won't exceed $m - 1$.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: does not exceed

Comment (ZamDimon) [reply lectures/zksnark.tex]: Also, isn't it always exactly $m-1$?

File: lectures/zksnark.tex [hunk 33]
```diff
@@ -0,0 +1,812 @@
...
+        &= -\frac{5}{6}x^3 + 6x^2 - \frac{79}{6}x + 9
+    \end{align*}
+    Therefore, the final Lagrange interpolation polynomial is:
+    \begin{equation*}
+        A_1(x) = -\frac{5}{6}x^3 + 6x^2 - \frac{79}{6}x + 9
+    \end{equation*}
+
+    As shown in Figure~\ref{fig:example-interpolation}, the curve intersects all the given points.
+    In this figure, the x-axis represents the constraint number, and the y-axis represents the 
+    coefficients of the $x_1$ пwitness element.
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Imposter п

File: lectures/zksnark.tex [hunk 34]
```diff
@@ -0,0 +1,812 @@
...
+        
+        \end{axis}
+    \end{tikzpicture}
+    \caption{The Lagrange inteprolation polynomial for points $\{(1,1), (2,0), (3,1), (4,-1)\}$}
+    \label{fig:example-interpolation}
+\end{figure}
+
+The degree of coefficient polynomials won't exceed $m - 1$.
+
+Now, using coefficients encoded with polynomials, a constraint number $X, X \in \{1, \dots\ m\}$, from
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: ...a constraint number $X \in \\{1,\dots,m\\}$, from...

File: lectures/zksnark.tex [hunk 35]
```diff
@@ -0,0 +1,812 @@
...
+a constraint system with a witness vector $w$ can be built in the next way:
+\begin{align*}
+    (w_1A_1(X) + w_2A_2(X) + \dots + w_nA_n(X)) &\times (w_1B_1(X) + w_2B_2(X) + \dots + w_nB_n(X)) =\\ = (w_1C_1(X) + w_2C_2(X)& + \dots + w_nC_n(X))
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Here, you use capital letters to represent polynomials :)

File: lectures/zksnark.tex [hunk 36]
```diff
@@ -0,0 +1,812 @@
...
+\end{align*}
+Or:
+\begin{align*}
+    (\sum_{i = 1}^{n} w_iA_i(X)) \times (\sum_{i = 1}^{n} w_iB_i(X)) = (\sum_{i = 1}^{n} w_iC_i(X))
```
Comment (ZamDimon) [inline lectures/zksnark.tex]: Here, use `\left( ... \right)` for big brackets

**Review by ZamDimon — COMMENTED** [review]
All is good! Just sending small notes and possible corrections to your lecture notes.

File: lectures/zksnark.tex (hunk 22 above)
Comment (OmegaTymbJIep) [reply lectures/zksnark.tex]: Will it be okay as written below?

The inner product of a linear space $\mathbb{V}$ is any symmetric, linear in the first argument and positive binary function from the vector space to a set of scalars.

Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An **inner product** on $\mathbb{V}$ is a function:

$$
\langle \cdot, \cdot \rangle: \mathbb{V} \times \mathbb{V} \rightarrow \mathbb{F}
$$

defined for $u, v \in \mathbb{V}$ as follows:

$$
\langle u, v \rangle = \sum_{i=1}^{n} u_i v_i
$$

The basic properties that must be satisfied for all $c \in \mathbb{F}$ and for all $u, v, w \in \mathbb{V}$ are:

- **Symmetry:** $\langle u, v \rangle = \langle v, u \rangle$
- **Linearity in the first argument:** $\langle cu + v, w \rangle = c \langle u, w \rangle + \langle v, w \rangle$
- **Positivity:** $\langle u, u \rangle \geq 0$ and $\langle u, u \rangle = 0 \iff u = 0$

File: lectures/zksnark.tex (hunk 33 above)
Comment (OmegaTymbJIep) [reply lectures/zksnark.tex]: image

File: lectures/zksnark.tex (hunk 32 above)
Comment (OmegaTymbJIep) [reply lectures/zksnark.tex]: e.g. for points ${(0, 0), (1, 1), (2, 1), (3, 0)\}$ we can build a polynomial $p(x) = -\frac{1}{2}x^2 + \frac{3}{2}x$

Comment (ZamDimon) [reply lectures/zksnark.tex]: Yeah, fair enough. Even more trivial example is '{(1,1),(2,2),(3,3),(4,4)}', corresponding to 'f(x)=x'

**Review by ZamDimon — APPROVED** [review]
LGTM. The live speech was awesome! :)
